jupyterhub/tests/selenium/test_browser.py

```diff
@@ -209,36 +225,686 @@ async def test_open_url_login(
+    next_url = url_path_join(public_host(app), app.base_url, '/hub/home')
+    await in_thread(browser.get, next_url)
+
+    buttons = browser.find_elements(*HomePageLocators.BUTTONS_SERVER)
+    if len(buttons) != 2:
+        await webdriver_wait(browser, EC.visibility_of_all_elements_located(buttons))
+
+    u = app.users[orm.User.find(app.db, user)]
+    while not u.spawner.ready:
+        await asyncio.sleep(0.1)
+    """add this stop click event is registred in JS to verify that the poccess is not still pending"""
+    # Stop server via clicking on the Stop button
+    click(browser, HomePageLocators.BUTTON_STOP_SERVER)
+    # verify that the stop button is invisible on the page
+    await webdriver_wait(
+        browser, EC.invisibility_of_element_located(HomePageLocators.BUTTON_STOP_SERVER)
+    )
+    buttons = browser.find_elements(*HomePageLocators.BUTTONS_SERVER)
+    assert len(buttons) == 2
+    # checking attributes of buttons after server was stopped
+    for button in buttons:
+        id = button.get_attribute('id')
+        href = button.get_attribute('href')
+        style = button.get_attribute('style')
+    assert buttons[0].get_attribute('id') == "stop"
+    assert buttons[0].get_attribute('style') == "display: none;"
+    assert buttons[0].get_attribute('href') == None
+    assert buttons[1].get_attribute('id') == "start"
+    assert buttons[1].get_attribute('href').endswith(f"/hub/spawn/{user}")
+    assert buttons[1].text == HomePageLocators.BUTTON_START_SERVER_NAME_DOWN
 
 
 # TOKEN PAGE
-async def open_token_page(app, browser, user="test_user", pass_w="test_user"):
+async def open_token_page(app, browser, user, pass_w):
+    """function to open the token page"""
 
-    url = url_path_join(public_host(app), app.hub.base_url, "/login?next=/hub/token")
-    await in_thread(browser.get, url)
-    redirected_url = url_path_join(public_host(app), app.base_url, '/hub/token')
+    token_page = url_escape(app.base_url) + "hub/token"
+    await open_url(app, browser, param="/login?next=" + token_page)
+    await login(browser, user, pass_w)
+    while '/hub/token' not in browser.current_url:
+        await asyncio.sleep(0.5)
+
+
+def elements_table(browser):
+    return browser.find_element(*TokenPageLocators.TABLE_API)
+
+
+def elements_table_body(browser):
+    table = browser.find_element(*TokenPageLocators.TABLE_API)
+    return table.find_element(*TokenPageLocators.TABLE_API_BODY)
+
+
+def elements_row_body(browser):
+    table = browser.find_element(*TokenPageLocators.TABLE_API)
+    body = table.find_element(*TokenPageLocators.TABLE_API_BODY)
+    return body.find_elements(*TokenPageLocators.TABLE_API_ROWS_BY_CLASS)
+
+
+def elements_row_head(browser):
+    table = browser.find_element(*TokenPageLocators.TABLE_API)
+    head = table.find_element(*TokenPageLocators.TABLE_API_HEAD)
+    return head.find_elements(*TokenPageLocators.TABLE_API_ROWS)
+
+
+async def test_elements_of_token_page_server_not_started(
```

Comment:
token page isn't really sensitive to server running, instead let's create a couple of tokens with e.g. `user.new_api_token` and check the page contents. We don't need two separate tests.